### PR TITLE
[restriction_policy] Add restriction policy resource

### PR DESCRIPTION
**Pull request by agnesjang98**
Adds restriction policy resource to terraform provider

example terraform apply
```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # datadog_restriction_policy.foo will be created
  + resource "datadog_restriction_policy" "foo" {
      + id          = (known after apply)
      + resource_id = "notebook:abc-def-ghi"

      + bindings {
          + principals = [
              + "role:61dcfc2c-eee2-11ed-9181-da7ad0900002",
              + "role:7ac320d0-1486-11ee-a488-da7ad0900002",
            ]
          + relation   = "editor"
        }
      + bindings {
          + principals = [
              + "org:1376fc06-92ba-11ed-8615-da7ad0900002",
            ]
          + relation   = "viewer"
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

datadog_restriction_policy.foo: Creating...
datadog_restriction_policy.foo: Creation complete after 0s [id=notebook:abc-def-ghi]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

### CODE REVIEW

File: datadog/fwprovider/resource_datadog_restriction_policy.go
Comment (sanjaythakkar-dd) [inline datadog/fwprovider/resource_datadog_restriction_policy.go:220]: what kind of error is thrown if formatting/input value is incorrect?

File: docs/resources/restriction_policy.md
Comment (sanjaythakkar-dd) [inline docs/resources/restriction_policy.md:47]: since we are only mentioning role and org, do you think we would need a FF in the backend to block teams/users temporarily intil that feature is GA'ed?

File: docs/resources/restriction_policy.md
Comment (sanjaythakkar-dd) [inline docs/resources/restriction_policy.md:1]: do you think we would need Sid's input on this documentation?

**Review by sanjaythakkar-dd — COMMENTED** [review]
good work on the PR @agnesjang98 
can we also have descrption for the PR please

**Review by skarimo — DISMISSED** [review]

File: datadog/fwprovider/resource_datadog_restriction_policy.go
Comment (agnesjang98) [reply datadog/fwprovider/resource_datadog_restriction_policy.go:220]:
```
400 Bad Request: {"errors":["Invalid Request:
        org:uuid must be able to schedule_report,
        share and view"]}
```
consistent w api i believe

```
 400 Bad Request: {"errors":["Invalid request: ID parsed from principal 'role:bad-uuid' is invalid, 'bad-uuid' is not a valid UUID"]}
```

File: docs/resources/restriction_policy.md
Comment (agnesjang98) [reply docs/resources/restriction_policy.md:47]: not gating teams, but not documenting yet so we'll leave this as is

File: docs/resources/restriction_policy.md
Comment (agnesjang98) [reply docs/resources/restriction_policy.md:1]: this was autogenerated and the params documentation are consistent with the API, can ping him if you still feel like it needs a look

File: datadog/fwprovider/resource_datadog_restriction_policy.go
Comment (sanjaythakkar-dd) [reply datadog/fwprovider/resource_datadog_restriction_policy.go:220]: little confused, naive question but if we are directly calling zoltron apis w/o any formatting check, the errors should be same right?

File: datadog/fwprovider/resource_datadog_restriction_policy.go
Comment (sanjaythakkar-dd) [inline datadog/fwprovider/resource_datadog_restriction_policy.go:222]: if `bindings` is not provided or is empty, what would be the expected behavior ?

File: datadog/tests/resource_datadog_restriction_policy_test.go
Comment (sanjaythakkar-dd) [inline datadog/tests/resource_datadog_restriction_policy_test.go:1]: what are your thoughts on adding a few more test scenarios
updating RP
empty bindings
invalid input format

File: docs/resources/restriction_policy.md
Comment (sanjaythakkar-dd) [reply docs/resources/restriction_policy.md:1]: discussed on slack, Agnes will get f/b from Sid

**Comment by sanjaythakkar-dd** [comment]
what are your thoughts on adding example in PR description for how it looks like when we do terraform apply

File: datadog/fwprovider/resource_datadog_restriction_policy.go
Comment (agnesjang98) [reply datadog/fwprovider/resource_datadog_restriction_policy.go:220]: yeah they should be the same

File: docs/resources/restriction_policy.md
Comment (agnesjang98) [reply docs/resources/restriction_policy.md:41]: what additional details would you want here? will test that out but should be consistent with our API because the principals and relation are marked as optional in the resource go file

Comment (sanjaythakkar-dd) [reply docs/resources/restriction_policy.md:41]: actually nvm, i saw that you added the details in [this](https://github.com/DataDog/terraform-provider-datadog/blob/79824a62707ac276117d297bed68a09027500546/docs/resources/restriction_policy.md#nested-schema-for-bindings) section

File: datadog/fwprovider/resource_datadog_restriction_policy.go
Comment (sanjaythakkar-dd) [reply datadog/fwprovider/resource_datadog_restriction_policy.go:220]: ah then thats good. thank you!

File: datadog/fwprovider/resource_datadog_restriction_policy.go
Comment (agnesjang98) [reply datadog/fwprovider/resource_datadog_restriction_policy.go:222]: ```400 Bad Request: {"errors":["Invalid request: update requests must contain at least one binding."]}```

Comment (agnesjang98) [reply datadog/fwprovider/resource_datadog_restriction_policy.go:222]: ```400 Bad Request: {"errors":["Invalid request: principals list cannot be empty; found empty list for relation ''"]}```
^ if i do `bindings {}`

Comment (agnesjang98) [reply datadog/fwprovider/resource_datadog_restriction_policy.go:222]: If you want to delete the restriction policy, it'd have to be like this 
```terraform destroy -target='datadog_restriction_policy.foo' ```

**Comment by sanjaythakkar-dd** [comment]
good work here @agnesjang98 
Here is some data from my local testing with this PR

**success scenarios:**
```
* single principal for relation editor ✔️ 
* multiple principals for relation editor ✔️ 
* multiple bindings with viewer and editor relations ✔️ 
     * note: ordering is imp, in the list of bindings, cannot set `org` as `viewer` at index 0 and then setting `role` as `editor`. This will result in an error
```
**error scenarios:**

```
* setting empty bindings list
│ 400 Bad Request: {"errors":["Invalid request: principals list cannot be empty; found empty list for relation ''"]}

* no bindings provided
│ 400 Bad Request: {"errors":["Invalid request: update requests must contain at least one binding."]}

* setting viewer as relation for org & role for dashboard
│ 400 Bad Request: {"errors":["Invalid request: principal type 'role' is not supported"]}

* using resource-type as dashboard2
│ 400 Bad Request: {"errors":["Invalid resource type: 'dashboard2' does not support restriction policies"]}

* duplicating same binding block
│ 400 Bad Request: {"errors":["Invalid request: cannot contain duplicate relations; found duplicate 'viewer'"]}

* setting  org and roles as `viewer`
│ 400 Bad Request: {"errors":["Invalid request: principal type 'role' is not supported"]}

* using same principal for viewer and editor
│ 400 Bad Request: {"errors":["Invalid request: a principal can only have one relation; found 'role:7ac320d0-1486-11ee-a488-da7ad0900002' in two or more bindings"]}

* using keyword roles instead of role
│ 400 Bad Request: {"errors":["Invalid request: principal type 'roles' is not supported"]}

*invalid role uuid
│ 400 Bad Request: {"errors":["Invalid Request: one or more roles do not exist in organization"]}

* incorrect length of role uuid
│ 400 Bad Request: {"errors":["Invalid request: ID parsed from principal 'role:x61dcfc2c-eee2-11ed-9181-da7ad0900002' is invalid – it must not be longer than 36 characters"]}

* forgetting the : between principal type and principal id
│ 400 Bad Request: {"errors":["Invalid request: the provided principal 'role61dcfc2c-eee2-11ed-9181-da7ad0900002' is not in the correct format. Please provide the ID in the format 'type:id'"]}
```

Interesting case
```
* set org as viewer at index 0 and then role as editor at index 1

datadog_restriction_policy.foo: Creating...
╷
│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to datadog_restriction_policy.foo, provider "provider[\"registry.terraform.io/datadog/datadog\"]" produced an unexpected new value: .bindings[0].principals[0]: was cty.StringVal("org:1376fc06-92ba-11ed-8615-da7ad0900002"), but now cty.StringVal("role:61dcfc2c-eee2-11ed-9181-da7ad0900002").
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
╷
│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to datadog_restriction_policy.foo, provider "provider[\"registry.terraform.io/datadog/datadog\"]" produced an unexpected new value: .bindings[0].principals: new element 1 has appeared.
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
╷
│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to datadog_restriction_policy.foo, provider "provider[\"registry.terraform.io/datadog/datadog\"]" produced an unexpected new value: .bindings[0].relation: was cty.StringVal("viewer"), but now cty.StringVal("editor").
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
╷
│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to datadog_restriction_policy.foo, provider "provider[\"registry.terraform.io/datadog/datadog\"]" produced an unexpected new value: .bindings[1].principals[0]: was cty.StringVal("role:61dcfc2c-eee2-11ed-9181-da7ad0900002"), but now cty.StringVal("org:1376fc06-92ba-11ed-8615-da7ad0900002").
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
╷
│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to datadog_restriction_policy.foo, provider "provider[\"registry.terraform.io/datadog/datadog\"]" produced an unexpected new value: .bindings[1].principals: element 1 has vanished.
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
╷
│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to datadog_restriction_policy.foo, provider "provider[\"registry.terraform.io/datadog/datadog\"]" produced an unexpected new value: .bindings[1].relation: was cty.StringVal("editor"), but now cty.StringVal("viewer").
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.

```

**Comment by sanjaythakkar-dd** [comment]
`terraform plan`
```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # datadog_restriction_policy.foo will be created
  + resource "datadog_restriction_policy" "foo" {
      + id          = (known after apply)
      + resource_id = "notebook:abc-def-ghi"

      + bindings {
          + principals = [
              + "role:61dcfc2c-eee2-11ed-9181-da7ad0900002",
              + "role:7ac320d0-1486-11ee-a488-da7ad0900002",
            ]
          + relation   = "editor"
        }
      + bindings {
          + principals = [
              + "org:1376fc06-92ba-11ed-8615-da7ad0900002",
            ]
          + relation   = "viewer"
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```

`terraform apply`
```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # datadog_restriction_policy.foo will be created
  + resource "datadog_restriction_policy" "foo" {
      + id          = (known after apply)
      + resource_id = "notebook:abc-def-ghi"

      + bindings {
          + principals = [
              + "role:61dcfc2c-eee2-11ed-9181-da7ad0900002",
              + "role:7ac320d0-1486-11ee-a488-da7ad0900002",
            ]
          + relation   = "editor"
        }
      + bindings {
          + principals = [
              + "org:1376fc06-92ba-11ed-8615-da7ad0900002",
            ]
          + relation   = "viewer"
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

datadog_restriction_policy.foo: Creating...
datadog_restriction_policy.foo: Creation complete after 0s [id=notebook:abc-def-ghi]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

File: examples/resources/datadog_restriction_policy/resource.tf
Comment (agnesjang98) [inline examples/resources/datadog_restriction_policy/resource.tf:6]: @skarimo these are autogenerated right? wondering why resource_id isn't getting added here after I made it required

File: docs/resources/restriction_policy.md
Comment (agnesjang98) [inline docs/resources/restriction_policy.md:20]: @skarimo - same thing here, these are autogenerated right? also wondering why resource_id isn't getting added here after I made it required

File: datadog/fwprovider/resource_datadog_restriction_policy.go
Comment (agnesjang98) [reply datadog/fwprovider/resource_datadog_restriction_policy.go:24]: not sure what the implications are but this file was autogenerated, I would want to keep everything consistent with other resource files generated by the provider
https://github.com/DataDog/terraform-provider-datadog/blob/master/datadog/fwprovider/resource_datadog_team.go#L21-L24

File: datadog/fwprovider/resource_datadog_restriction_policy.go
Comment (agnesjang98) [reply datadog/fwprovider/resource_datadog_restriction_policy.go:44]: https://github.com/DataDog/terraform-provider-datadog/blob/master/datadog/fwprovider/resource_datadog_team.go#L41 this seems to be no longer relevant, removing the second empty assignment

File: datadog/fwprovider/resource_datadog_restriction_policy.go
Comment (vincentyl) [reply datadog/fwprovider/resource_datadog_restriction_policy.go:24]: hmm ok

File: examples/resources/datadog_restriction_policy/resource.tf
Comment (skarimo) [reply examples/resources/datadog_restriction_policy/resource.tf:6]: The generator is ran once to generate the boiler code. Any changes afterwards will need to be done manually

File: docs/resources/restriction_policy.md
Comment (skarimo) [reply docs/resources/restriction_policy.md:20]: This is generated from the file: [examples/resources/datadog_restriction_policy/resource.tf](https://github.com/DataDog/terraform-provider-datadog/pull/1957/files/ccbe98b9b8919e4c94b5ebdb05931a17fd784d82#diff-76b94324cb4459196e393222837a2e0a992d09af0541b6ca21cb5c0cee57cabc)

So making changes to that and running `make docs` will reflect the changes here

File: datadog/fwprovider/resource_datadog_restriction_policy.go
Comment (agnesjang98) [reply datadog/fwprovider/resource_datadog_restriction_policy.go:53]: updated to use a set @sanjaythakkar-dd  if you wanna try running that test again

File: datadog/fwprovider/resource_datadog_restriction_policy.go
Comment (vincentyl) [inline datadog/fwprovider/resource_datadog_restriction_policy.go:204]: just curious, if the principals is empty, shall we proceed this logic?

Comment (vincentyl) [reply datadog/fwprovider/resource_datadog_restriction_policy.go:204]: I mean adding bindingsTfItem to the Bindings

Comment (agnesjang98) [reply datadog/fwprovider/resource_datadog_restriction_policy.go:204]: We have validation in place to prevent empty principal lists.
Example call:
```
resource "datadog_restriction_policy" "foo" {
  resource_id = "notebook:id"
  bindings {
    principals = []
    relation = "editor"
  }
}
```

Result:
```
Error: error retrieving RestrictionPolicy 
with datadog_restriction_policy.foo,
on main.tf line 14, in resource "datadog_restriction_policy" "foo":
14: resource "datadog_restriction_policy" "foo" {

400 Bad Request: {"errors":["Invalid request: principals list cannot be empty; found empty list for relation 'editor'"]}
```

Comment (vincentyl) [reply datadog/fwprovider/resource_datadog_restriction_policy.go:204]: I see

**Review by vincentyl — APPROVED** [review]
my knowledge of terraform plugin is very limited. please get sanjay's approval before merging

File: datadog/fwprovider/resource_datadog_restriction_policy.go
Comment (sanjaythakkar-dd) [reply datadog/fwprovider/resource_datadog_restriction_policy.go:53]: good point @anton-dd 
looks good now @agnesjang98 , good work

**Review by sanjaythakkar-dd — APPROVED** [review]
LGTM

**Review by skarimo — APPROVED** [review]

**Pull request merged**
